### PR TITLE
2 get topics

### PR DESCRIPTION
**Pull request by nickdip**
added invalid path url

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,3 +1,11 @@
 # Northcoders News API
 
-For instructions, please head over to [L2C NC News](https://l2c.northcoders.com/courses/be/nc-news).
+## Creating environment variables:
+This project uses the [dotenv](https://www.npmjs.com/package/dotenv) package and requires a .env.development file and .env.test file in the main directory. 
+
+.env.test requires PGDATABASE=nc_news_test to access the test database
+.env.developer requires PGDATABASE=nc_news to access the developement data
```
Comment (itskatherine) [inline README.md:7]: Nice clear instructions , though should be .env.development (not developer) 🔍

File: controllers/topics.controllers.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+const { fetchTopics } = require("../models/topics.models")
+
+
+exports.getTopics = (req, res, next) => {
+    fetchTopics()
+    .then( (result) => {
+        if (result.status === 200) res.status(200).send( { msg: result.topics })
```
Comment (itskatherine) [inline controllers/topics.controllers.js:7]: Returning the topics on the key of `msg` is a bit of a confusing naming convention - something like `topics` would be preferable.

File: models/topics.models.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+const db = require('../db/connection')
+
+exports.fetchTopics = (req, res, next) => {
+    return db.query('SELECT * FROM topics;').then(( { rows } ) => {
+        if (!rows.length) return { status: 404, msg: 'No topics found' }
```
Comment (itskatherine) [inline models/topics.models.js:5]: No need for this error at the moment, and you don't have a test for it
If there were no topics in the database, it wouldn't result in a 404 error - it would still be a 200 but returning an empty array

File: models/topics.models.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+        return { status: 200, topics: rows}
+     })
+     .catch((err) => {
+        if (err.code === "42P01") return Promise.reject({code: "42P01", msg: 'Database Error: Table does not exist'})
```
Comment (itskatherine) [inline models/topics.models.js:9]: Same with this error, no test for it and it isn't needed 
When the user has information they need to input (such as an id they include in the url), these SQL errors can occur.
Really the only thing that can cause an error with this endpoint would be if they spelled the endpoint incorrectly, which you've tested for well 👍

File: models/topics.models.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+        return Promise.reject({status: 500, msg: 'Internal Server Error'})
```
Comment (itskatherine) [inline models/topics.models.js:10]: There's no need to prompt a 500 server error, it's something that's caused by things out of your control, like the server going down

File: app.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+const express = require("express")
+const app = express()
+
+const controllers = require("./controllers/index.controllers")
+
+app.use((req, res, next) => {
+    validPaths = ["/api/topics"]
+    if (!validPaths.includes(req.path)) res.status(404).send({ msg: "Path not found" })
+    //console.log("HELLO")
```
Comment (itskatherine) [inline app.js:9]: Should get rid of console.logs or commented out code for PRs

File: app.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    next()
+})
```
Comment (itskatherine) [inline app.js:11]: This approach works, though could achieve the same thing with a `app.all("*", (req,res)=>{...send a 404 here})` after your endpoints

File: __tests__/app.test.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+describe("GET /api/topics", () => {
+    test("status:200, responds with an array of topic objects", () => {
+        return request(app)
+        .get("/api/topics")
+        .expect(200)
+        .then(({ body: { msg } }) => {
+            expect(msg).toHaveLength(3)
+            expect(msg[0].description).toEqual('The man, the Mitch, the legend')
+            expect(msg[0].slug).toEqual('mitch')
```
Comment (itskatherine) [inline __tests__/app.test.js:32]: A good test - ideally we would be looping through the array to check all the objects in it match the shape we're looking for, i.e. have the correct keys

File: __tests__/app.test.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+describe("Invalid", () => {
+    test("invalid path", () => {
+        return request(app)
+            .get("/api/topix")
+            .expect(404)
+            .then(({ body: { msg } }) => {
+                expect(msg).toBe("Path not found")
+            })
+        })
+
```
Comment (itskatherine) [inline __tests__/app.test.js:21]: Great test, I like that you've separated it from the endpoint to make it clear this will apply to any bad endpoint

File: controllers/errors.controllers.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+exports.handleCustomErrors = (err, req, res, next) => {
+    console.log("HELLO")
+    console.log(err)
+    if (err.status = 404) return res.status(404).send({ msg: err.msg })
+    next(err)
+
+}
+
+exports.handlePSQLErrors = (err, req, res, next) => {
```
Comment (itskatherine) [inline controllers/errors.controllers.js:11]: This endpoint isn't using this`handlePSQLErrors` error handling function (or the `handleCustomErrors` one), so ideally would not be included in the PR for this ticket. 
It's important to limit code to what is required for the ticket.

File: laterfiles.js [hunk 11]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (itskatherine) [inline laterfiles.js:1]: laterfiles should be in a later PR 🙏

**Review by itskatherine — COMMENTED** [review]
Hi Nick, some nice work - the main things to look out for in this PR are to do with including some code and error handling that doesn't apply to this ticket - which I've pointed out below
It's important to only include code in a ticket that meet the needs of that specific ticket
No need to resubmit but please have a look through my comments before merging

File: controllers/topics.controllers.js (hunk 2 above)
Comment (nickdip) [reply controllers/topics.controllers.js:7]: Oh thanks yes. It's because I had other tests but kev already asked me to remove them!
